Refactor the Argument Editor

camielvs opened the pull request:
Closes https://github.com/Cloud-Pipelines/pipeline-studio-app/issues/148
Closes https://github.com/Cloud-Pipelines/pipeline-studio-app/issues/149
Closes https://github.com/Shopify/oasis-frontend/issues/29

Refactors the componentry around the Argument Editor to achieve the following:

- Unstack ArgumentEditorDialog, ArgumentEditor and ArgumentInput into individual files for a more clear separation of concerns and simplification of props
- Decouple the Draggable Dialog from the Argument Editor Dialog, and standardise the Draggable Dialog into a generic component that can be utilised elsewhere in the app
- Combine the InputSpec and Argument Data into a singular ArgumentInput array that keeps track of the current state of argument input data within the editor. The array also keeps fields in the same order throughout the editing experience
- Input boxes now have a "Undo Changes" button which reverts the input state back to what it was when the editor was opened.
- Dedicated "Include/exclude argument" button to add/remove an argument from the TaskSpec when submitted. When an argument is added it will be assigned its default value if a value does not already exist.
- New "Reset to Default" button to reset the input value to the defined default.
- "Undefined" and "Empty String" are now correctly tracked and treated differently by the system.
- Required arguments will highlight red when "Excluded"
- Fixes a React error regarding controlled/uncontrolled input components (empty values must always be empty strings, but are now correctly handled as needed when submitted to the Spec via the "Apply" button)
- Fixes a bug in the Draggable Dialog that would increment its z-index every time it was clicked, even if it was already on top of all other dialogs
- Argument Editor Dialog will now open with its position relative to the node that triggered it
- Clearer component flow and more readable code!

User interface itself now has a few new buttons:

- UNDO CHANGES - reverts any changes made to the field so it returns to the state it was in when the editor was first opened.
- EXCLUDE ARGUMENT - disables the argument (it will grey out), removing it from the task spec when edits are applied [aka "undefined"].
- INCLUDE ARGUMENT - enables the argument, including it in the task spec when edits are applied.
- RESET TO DEFAULT - changes the argument value to whatever the default it. For arguments connected to another node this will sever the connection.

![image](https://github.com/user-attachments/assets/56d077a6-7a75-48fb-b257-817bf1651559)

Demo: https://screenshot.click/01-44-68m0h-e2hc1.mp4

camielvs commented on src/DragNDrop/DraggableDialog.tsx:
`DraggableDialog` extracted and split into its own component from `ArgumentEditorDialog`

Ark-kun commented on src/components/ArgumentEditor/ArgumentInputField.tsx:
```diff
@@ -12,78 +12,38 @@ import { Button } from "@/components/ui/button";
     setInputValue(value);
   };
 
   const handleRemove = () => {
```
Are you sure this code has correct behavior?

Correct behavior of "Reset" button:
* If `input_spec.optional` then remove the argument completely from `task_spec.arguments`
* Otherwise if `input_spec.default` set the argument value to `str(input_spec.default)`
* Otherwise set the argument value to `""`

For example, if `.default` has value, the "Reset" button should still remove the argument completely.

I do not see handling of `.optional`.

camielvs replied:
Empty optional arguments are removed when the "Apply" button is clicked and not beforehand. I've verified via that empty optional arguments do not end up in the task spec.

For context:
The Editor now runs on its own `ArgumentInput[]` props which keeps track of the editor state. Within this state the `optional` arguments exist (because they need to be displayed on screen). However, when the State is submitted to the Spec the blank optional arguments are removed. See here for where the empty arguments are handled on submission: https://github.com/Cloud-Pipelines/pipeline-studio-app/pull/163/files#diff-bdff881229b38628cd16f0ee616a399ef696f83b95a533965a1798b94682ea49R66

camielvs commented:
Identified a bug where the onDragEnd event is not firing for the Draggable Dialog. Will investigate, fix & reopen.

camielvs commented:
> Identified a bug where the onDragEnd event is not firing for the Draggable Dialog. Will investigate, fix & reopen.

Fixed.

Also fixed a bug where the z-index of a dialog would continuously increase every time it was clicked.

Ark-kun commented on src/DragNDrop/DraggableDialog.tsx:
```diff
@@ -0,0 +1,152 @@
+  id,
+  title,
+  description,
+  position = { x: 100, y: 100 },
```
In the future, consider positioning the dialog based on the task node position.

camielvs replied:
This is just copied over from the existing code for the original Draggable Dialog component. I've updated it now to be based on node position.

Ark-kun commented on src/componentSpec.ts:
```diff
@@ -199,6 +199,17 @@ export interface TaskOutputArgument {
+  key: string;
+  value: ArgumentType;
+  initialValue: ArgumentType;
+  defines: InputSpec;
```
`inputSpec`?

camielvs replied:
This is a type

Mbeaulne replied:
NIT: I will leave it up to you, but maybe we should rename it from defines to InputSpec if that is the type name? At first glance, I don't _really_ understand defines.

camielvs replied:
That's a fair comment. I've changed it.

Ark-kun commented on src/components/ArgumentsEditor/ArgumentsEditorDialog.tsx:
```diff
@@ -0,0 +1,107 @@
+    <DraggableDialog
+      id="arguments-dialog"
+      title={componentSpec.name ?? "<component>"}
+      description={`Configure the component's input parameters.`}
```
"input parameters" -> "input arguments".

Ark-kun commented on src/components/ArgumentsEditor/ArgumentsEditorDialog.tsx:
```diff
@@ -0,0 +1,107 @@
+    return null;
+  }
+
+  const handleApply = () => {
```
Could you please create test cases that verify the following:

Optional inputs:
* Possible for user to specify empty string argument value "": `arguments: {"input1": ""}`
* Possible for user to specify argument value that equals the default value: `arguments: {"input1": "some_default"}`
* Resetting the argument removes the key regardless of `.default` string argument value "": `arguments: {}`

Non-optional inputs:
* Possible for user to specify empty string argument value "": `arguments: {"input1": ""}`
* Possible for user to specify argument value that equals the default value: `arguments: {"input1": "some_default"}`
* Resetting the argument sets the value to the `input_spec.default` value if `input_spec.default` is defined: `arguments: {"input1": "some_default"}`
* Resetting the argument sets the value to empty string if `input_spec.default` is undefined: `arguments: {"input1": ""}`

camielvs replied:
As noted in today's conversion we will defer tests until later.

Mbeaulne reviewed (changes requested):
In the current version, clicking the argument dialog would bring it to the front of the stack. I think this is important, especially on smaller screens or with bigger dialogus

https://github.com/user-attachments/assets/55b0e279-850c-4934-84f2-982e4ed06d15

Mbeaulne commented on src/DragNDrop/DraggableDialog.tsx:
if we are creating a new component, lets move it into the components folder. I plan on moving everything out of dragNDrop at some point

camielvs replied:
Agree - though for tidiness I think it might be easier to move everything at once to a "shared" folder in a separate PR.

camielvs commented on src/DragNDrop/DraggableDialog.tsx:
```diff
@@ -0,0 +1,152 @@
+      return; // Already on top
+    }
+
+    const newZIndex = globalZIndexCounter.increment();
```
Huuuh. This was working yesterday when I requested reviews. Strange. I have looked at it again and it should definitely be working now.

Mbeaulne commented on src/components/ArgumentsEditor/ArgumentsEditorDialog.tsx:
```diff
@@ -0,0 +1,114 @@
+    <DraggableDialog
+      id="arguments-dialog"
+      title={componentSpec.name ?? "<component>"}
+      description={`Configure the component's input arguments.`}
```
```suggestion
      description="Configure the component's input arguments."
```

Mbeaulne commented on src/components/ArgumentsEditor/ArgumentInputField.tsx:
```diff
@@ -0,0 +1,241 @@
+  const handleInputChange = (value: string) => {
+    setInputValue(value);
+
+    const def = getDefaultValue(argument);
```
can we name this defaultValue or something? `def` is a little confusing at first glance

Mbeaulne replied:
On a side note, I don't mind short variable names like 

```
onClickHandler = (e: Event) => {}
```

As long as its a easily understood. However, I tend to be more explicit in my naming even if they feel a bit too long at times.

camielvs replied:
Yes, sure. I originally used `default` but it's a reserved variable name and my mind went shorter rather than longer. I agree `defaultValue` will actually be much clearer.

Mbeaulne commented on src/components/ArgumentsEditor/ArgumentInputField.tsx:
```diff
@@ -0,0 +1,241 @@
+  const [inputValue, setInputValue] = useState(getInputValue(argument));
+  const [hidePlaceholder, setHidePlaceholder] = useState(argument.linkedNode);
+
+  const undoValue = useMemo(() => argument, []);
```
This is actually a cool use of useMemo. I will have to remember it!

Mbeaulne commented on src/DragNDrop/ComponentTaskNode.tsx:
```diff
@@ -327,12 +330,32 @@ const ComponentTaskNode = ({ data }: NodeProps) => {
           closeEditor={closeArgumentsEditor}
           setArguments={typedData.setArguments}
           disabled={!!runStatus}
+          initialPosition={{
+            // On large screens position the dialog to the right of the node
+            // On small screens position the dialog below the node & center it
+            // Assumed 650px width for the dialog & 720px small screen breakpoint
+            x:
+              window.innerWidth <= 720 ||
+              nodeRef.current.getBoundingClientRect().right + 8 >
+                window.innerWidth - 650
+                ? nodeRef.current.getBoundingClientRect().left +
+                  nodeRef.current.getBoundingClientRect().width / 2 -
+                  650 / 2
+                : nodeRef.current.getBoundingClientRect().right + 8,
+
+            y:
+              window.innerWidth <= 720 ||
+              nodeRef.current.getBoundingClientRect().right + 8 >
+                window.innerWidth - 650
+                ? nodeRef.current.getBoundingClientRect().bottom + 8
+                : nodeRef.current.getBoundingClientRect().top,
+          }}
```
Lets move this logic out of the markup into a function or variable

camielvs replied:
Done!

Ark-kun commented on src/componentSpec.ts:
```diff
@@ -199,6 +199,19 @@ export interface TaskOutputArgument {
+/**
+ * Argument input to the Argument Editor.
+ */
+export type ArgumentInput = {
```
Please, let's move this type out of this file. This `componentSpec.ts` file is unrelated to UX. Let's not mix the UX types and core business logic types.
